src/v/redpanda/admin/debug.cc

```diff
@@ -229,29 +232,29 @@ void admin_server::register_debug_routes() {
       [this](std::unique_ptr<ss::http::request>) {
           vlog(adminlog.info, "Request to get leaders info");
           using result_t = ss::httpd::debug_json::leader_info;
-          std::vector<result_t> ans;
-
+          // TODO(rockwood): get_leaders can lead to a reactor stall, fix to use
+          // an async version
           auto leaders_info = _metadata_cache.local().get_leaders();
-          ans.reserve(leaders_info.size());
-          for (const auto& leader_info : leaders_info) {
-              result_t info;
-              info.ns = leader_info.tp_ns.ns;
-              info.topic = leader_info.tp_ns.tp;
-              info.partition_id = leader_info.pid;
-              info.leader = leader_info.current_leader.has_value()
-                              ? leader_info.current_leader.value()
-                              : -1;
-              info.previous_leader = leader_info.previous_leader.has_value()
-                                       ? leader_info.previous_leader.value()
-                                       : -1;
-              info.last_stable_leader_term
-                = leader_info.last_stable_leader_term;
-              info.update_term = leader_info.update_term;
-              info.partition_revision = leader_info.partition_revision;
-              ans.push_back(std::move(info));
-          }
-
-          return ss::make_ready_future<ss::json::json_return_type>(ans);
+          return ss::make_ready_future<ss::json::json_return_type>(
+            ss::json::stream_range_as_array(
+              admin::lw_shared_container(std::move(leaders_info)),
+              [](const auto& leader_info) {
+                  result_t info;
+                  info.ns = leader_info.tp_ns.ns;
+                  info.topic = leader_info.tp_ns.tp;
+                  info.partition_id = leader_info.pid;
+                  info.leader = leader_info.current_leader.has_value()
+                                  ? leader_info.current_leader.value()
+                                  : -1;
+                  info.previous_leader = leader_info.previous_leader.has_value()
+                                           ? leader_info.previous_leader.value()
+                                           : -1;
```

Comment:
Nitpick: It's not really part of the refactor, which is fine, but it would be nice to use `.value_or(-1)` in these two places.

Reply:
I'll fix in a followup, I am going to merge as the perf team was asking about the oversized allocs